zeebe/engine/src/main/java/io/camunda/zeebe/engine/scaling/redistribution/RedistributionBehavior.java

```diff
@@ -0,0 +1,149 @@
+import io.camunda.zeebe.engine.state.immutable.RoutingState;
+import io.camunda.zeebe.protocol.impl.record.value.scaling.RedistributionProgress;
+import io.camunda.zeebe.protocol.impl.record.value.scaling.RedistributionRecord;
+import io.camunda.zeebe.protocol.record.ValueType;
+import io.camunda.zeebe.protocol.record.intent.scaling.RedistributionIntent;
+import java.util.SequencedCollection;
+import org.agrona.collections.IntHashSet;
+
+public final class RedistributionBehavior {
+  private final TypedCommandWriter commandWriter;
+  private final StateWriter stateWriter;
+  private final CommandDistributionBehavior distributionBehavior;
+  private final RoutingState routingState;
+  private final ProcessingState processingState;
+  private final RedistributionState redistributionState;
+
+  public RedistributionBehavior(
+      final Writers writers,
+      final CommandDistributionBehavior distributionBehavior,
+      final ProcessingState processingState) {
+    commandWriter = writers.command();
+    stateWriter = writers.state();
+    this.distributionBehavior = distributionBehavior;
+    routingState = processingState.getRoutingState();
+    this.processingState = processingState;
+    redistributionState = processingState.getRedistributionState();
+  }
+
+  public void continueRedistribution(final long redistributionKey) {
+    // get current state and matching redistributor
+    final var stage = redistributionState.getStage();
+    final var progress = redistributionState.getProgress();
+    final var redistributor = redistributorForCurrentStage(stage);
+
+    // get all new redistributions
+    final var redistributions = redistributor.nextRedistributions(progress);
+
+    if (redistributions.isEmpty()) {
+      advanceToNextStage(redistributionKey, progress, stage);
+    } else {
+      advanceInSameStage(redistributionKey, progress, stage, redistributions);
+    }
+  }
+
+  private void advanceInSameStage(
+      final long redistributionKey,
+      final RedistributionProgress progress,
+      final RedistributionStage stage,
+      final SequencedCollection<? extends Redistribution<?>> redistributions) {
+    // find partitions to redistribute to
+    final var currentPartitions = routingState.currentPartitions();
+    final var desiredPartitions = routingState.desiredPartitions();
+    final var newPartitions = new IntHashSet(desiredPartitions.size() - currentPartitions.size());
+    newPartitions.addAll(desiredPartitions);
+    newPartitions.removeAll(currentPartitions);
+
+    // enqueue all redistributions
+    for (final var redistribution : redistributions) {
+      distributionBehavior
+          .withKey(redistribution.distributionKey())
+          .inQueue(DistributionQueue.REDISTRIBUTION)
+          .forPartitions(newPartitions)
+          .distribute(
+              redistribution.distributionValueType(),
+              redistribution.distributionIntent(),
+              redistribution.distributionValue());
+    }
```

Comment:
💭 It's probably rather unlikely but we should be cautious of the maximum record batch size here. Depending on the amount of partitions that we scale to and the amount of redistributions we could be writing quite a few events here. Since they have a queue they are likely small and it doesn't matter too much, but something to keep in mind anyways.